internal/controller/management/serviceaccounts/serviceaccounts.go

```diff
@@ -0,0 +1,301 @@
+func SetupReconcilerWithManager(kargoMgr manager.Manager, cfg ReconcilerConfig) error {
+	return ctrl.NewControllerManagedBy(kargoMgr).
+		For(&corev1.ServiceAccount{}).
+		WithEventFilter(
+			predicate.Funcs{
+				CreateFunc: func(event.CreateEvent) bool {
+					// Allow creation events to be handled for all ServiceAccounts
+					// This ensures any labeled or unlabeled SA gets proper
+					// reconciliation, including on restarts
+					return true
+				},
+				DeleteFunc: func(event.DeleteEvent) bool {
+					return false
+				},
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					oldSA, _ := e.ObjectOld.(*corev1.ServiceAccount)
+					newSA, _ := e.ObjectNew.(*corev1.ServiceAccount)
+					return hasControllerLabel(oldSA) != hasControllerLabel(newSA) ||
+						hasControllerLabel(newSA) && newSA.DeletionTimestamp != nil
+				},
+				GenericFunc: func(event.GenericEvent) bool {
+					return false
+				},
+			},
+		).
+		Complete(newReconciler(kargoMgr.GetClient(), cfg))
+}
+
+func newReconciler(kubeClient client.Client, cfg ReconcilerConfig) *reconciler {
+	r := &reconciler{
+		cfg:    cfg,
+		client: kubeClient,
+	}
+	r.getServiceAccountFn = r.client.Get
+	r.updateServiceAccountFn = r.client.Update
+	r.createRoleBindingFn = r.client.Create
+	r.updateRoleBindingFn = r.client.Update
+	r.deleteRoleBindingFn = r.client.Delete
+	r.listProjectFn = r.client.List
+	r.ensureControllerPermissionsFn = r.ensureControllerPermissions
+	r.removeControllerPermissionsFn = r.removeControllerPermissions
+	return r
+}
+
+// Reconcile is part of the main Kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"serviceAccount", req.NamespacedName.Name,
+	)
+	ctx = logging.ContextWithLogger(ctx, logger)
+	logger.Debug("reconciling ServiceAccount")
+
+	sa := &corev1.ServiceAccount{}
+	if err := r.getServiceAccountFn(ctx, req.NamespacedName, sa); err != nil {
+		if kubeerr.IsNotFound(err) {
+			logger.Debug("ServiceAccount not found, deleting RoleBindings for all Projects")
+			// If not found it means the ServiceAccount with the
+			// controller label was deleted, hence remove Controller
+			// Permissions for the ServiceAccount.
+			return ctrl.Result{}, r.removeControllerPermissionsFn(ctx, req.NamespacedName)
+		}
+		logger.Error(err, "Failed to get ServiceAccount")
+		return ctrl.Result{}, err
+	}
+
+	// Handle ServiceAccount deletion or updates where the
+	// controller label has been removed from the ServiceAccount.
+	// This indicates that the ServiceAccount is no longer managed by the controller,
+	// and we need to clean up any associated RoleBindings.
+	if sa.DeletionTimestamp != nil || !hasControllerLabel(sa) {
+		logger.Debug("Deleting RoleBindings for ServiceAccount", "serviceaccount", sa.Name)
```

Comment:
You already added this information to the logger on lines 135 - 137.

```suggestion
		logger.Debug("Deleting RoleBindings for ServiceAccount")
```